### PR TITLE
[Perf][TTS] omnivoice hot path: D2H batching, mask caching, cached text embeddings

### PR DESCRIPTION
**Pull request by JuanPZuluaga**
## Purpose

Cut per-step host overhead in the OmniVoice generator: batch the per-request GPU→CPU syncs,
cache the attention mask, and reuse cached text embeddings across steps instead of
recomputing them. OmniVoice's production deploy is `enforce_eager: true`, so this
host-overhead work is on the real serving path (not hidden inside a CUDA graph). Output
unchanged.

## Test Plan

Wall-clock A/B, branch vs base (origin/main), one RTX 6000 Ada, production eager deploy, 3
rounds/concurrency, all requests successful in both variants.

**vLLM Version:** 0.25.0

**vLLM-Omni Commit:** main

## Test Result

| concurrency | throughput (audio-s/s) | e2e p50 | e2e p99 |
|---|---|---|---|
| c=1  | 9.07 → 9.72 (**+7.2%**) | 886 → 828 ms (**−6.5%**) | 889 → 828 ms |
| c=8  | 8.96 → 9.37 (**+4.5%**) | 7157 → 6824 ms (−4.7%) | 7234 → 6962 ms |
| c=16 | 8.74 → 9.08 (**+3.9%**) | 14688 → 14128 ms (−3.8%) | 14812 → 14355 ms |

Positive and monotonic across all concurrencies; largest at c=1 where host overhead is the
biggest share of the step.

### CODE REVIEW

**Comment by JuanPZuluaga** [comment]
@hsliuustc0106 please let me know if there's anything else to be done here. 
thanks.

**Comment by linyueqian** [comment]
@vllm-omni-review-bot

**Comment by linyueqian** [comment]
A/B on an L20X box (vLLM 0.26.0, `omnivoice.yaml` as-is — `dtype: float32`, `enforce_eager: true`, auto-voice request). Baseline is `pr5174^2` = `8e44e780`, so the legs differ by exactly this PR. Two server instances per leg, first request after load discarded as cold:

| leg | warm samples | median |
|---|---|---|
| baseline `8e44e780` | 0.42 ×15 | **0.42s** |
| PR head `bb91c6e9` | 0.41 ×15 | **0.41s** |

Every warm sample on each side landed on the same value, across separate server starts — so the ~10ms gap is consistent, but it's also small (~2-3%) and my timer only resolves to 10ms, so read it as "a small real improvement", not a headline number. No regression, output size identical (203564 B) in all 30 requests, no failures.

**The TF32 path was not exercised**: `enable_tf32` defaults off (`OMNIVOICE_TF32` unset, not set in the yaml), so this measures the D2H batching, mask caching and cached text embeddings only. If you want the TF32 numbers I can re-run with `OMNIVOICE_TF32=1` on both legs — worth doing separately anyway since it's explicitly not bit-identical.

On the code: the float-mask switch looks consistent to me end-to-end — padding with `-inf` for float vs `False` for bool both mean "masked", and capturing the CUDA graph with a float dummy mask matches what `forward()` feeds at replay. Pre-doubling RoPE cos/sin is fine too; `_precompute_rope` has exactly one producer and `_apply_rotary_pos_emb` one consumer pair, both updated together, so there's no stale half-width reader left.

**Review by linyueqian — DISMISSED** [review]
Verified by A/B against `pr5174^2`: 0.41s vs 0.42s warm median, every one of 30 warm samples landing on the same value per leg across two server instances — small (~2-3%) but consistent, no regression, identical output size in all requests. Mask-dtype and RoPE pre-doubling read as consistent end-to-end (single producer/consumer pair, both updated together). Note the TF32 path is untested here since it defaults off.

**Comment by linyueqian** [comment]
I've dismissed my approval — CI caught something my e2e run didn't, and it's real. Sorry for the premature green.

`Engine&Model Executor Test` fails on this PR and passes on your other three (builds 13430/13431/13433 clean, 13432 red). Reproduced locally, and the baseline is a clean control:

```
# PR head 0d7d95b1
FAILED tests/model_executor/models/omnivoice/test_cuda_graph_generator.py::test_cuda_graph_disabled_matches_eager_generator
1 failed, 17 passed

# pr5174^2 (ee118ab4)
18 passed
```

The assertion is bit-exactness between the graph and eager paths:

```
torch.testing.assert_close(graph_logits, eager_logits, atol=0, rtol=0)
E  Mismatched elements: 935438 / 1049600 (89.1%)
E  Greatest absolute difference: 1.430511474609375e-06
```

The magnitude is floating-point noise, not broken math — so I don't think the optimisation is wrong. My guess is the mask change: the test feeds a **bool** `attn` mask, your `forward()` now materialises float from it, and the graph is captured with a **float** dummy mask, so capture and replay end up on different SDPA kernels and stop matching bit-for-bit. `atol=0, rtol=0` leaves no room for that.

Two ways out, and I don't have a strong preference:
- keep bit-exactness and make the captured mask path identical to what the test feeds, or
- if graph/eager can no longer be bit-identical by construction, loosen the test to a tight tolerance and say why in the test.

Either way it needs a decision from you rather than from me, since it's a question about what invariant the wrapper is supposed to guarantee. The perf result from my earlier comment still stands (~2-3%, no regression); this is purely about the equivalence test.

**Comment by linyueqian** [comment]
@JuanPZuluaga could you take a look at the graph/eager equivalence failure here? Details are in my comment above — short version is `test_cuda_graph_disabled_matches_eager_generator` asserts bit-exactness and this shifts results by ~1.4e-06, which I suspect is the bool-vs-float mask difference between what the test feeds and what the graph captures.

It needs your judgement rather than mine: either make the captured mask path match what the test feeds, or decide graph/eager can't be bit-identical any more and relax the assertion with a note. I re-approve as soon as CI is green — the perf side already checked out.

**Review by linyueqian — APPROVED** [review]
Re-approving — verified your fix on my side: `tests/model_executor/models/omnivoice` now goes 18 passed (was 1 failed / 17 passed on the previous head), and readthedocs is green on both PRs.

Nice touch documenting *why* a bool mask can't be cast straight into a float buffer in `_additive_float_mask` — True/False landing on 1.0/0.0 and silently unmasking is exactly the kind of thing that would be painful to find later.

One caveat so it's on the record: I couldn't re-measure perf on this head, because the speech endpoint on current main 500s for OmniVoice with `'OmniOpenAIServingSpeech' object has no attribute '_adapter'` — that's not yours (main at `pr5174^2` fails identically and your diff only touches the two omnivoice files), I filed it as #6118. So the ~2-3% from my earlier comment was measured on the pre-fix head; the mask conversion is small enough that I don't expect it to move, but I haven't proven that.
